source/geod24/concurrency.d

```diff
@@ -1097,59 +1133,124 @@ private:
 
         Start the scheduling loop
 
-        Param:
-            pos = m_fibers index of the Fiber to execute first
-
     ***********************************************************************/
 
-    void dispatch(size_t pos = 0)
+    void dispatch ()
     {
-        import std.algorithm.mutation : remove;
-
         thisScheduler(this);
         scope (exit) thisScheduler(null);
 
-        assert(pos < m_fibers.length);
-        while (m_fibers.length > 0)
+        MonoTime earliest_timeout;
+
+        while (!this.readyq.empty() || !this.wait_list.empty())
         {
-            // Is Fiber waiting on a FiberBinarySemaphore?
-            if (auto sem = m_fibers[pos].sem)
+            while (!readyq.empty())
             {
-                // Is condition met?
-                // TRUE: Clear the sem and schedule the fiber
-                // FALSE: Skip it
-                if (sem.shouldBlock())
+                InfoFiber cur_fiber = this.readyq.front();
+                this.readyq.removeFront();
+
+                assert(cur_fiber.state != Fiber.State.TERM);
+
+                auto t = cur_fiber.call(Fiber.Rethrow.no);
+
+                // Fibers that block on a FiberBinarySemaphore throw an
+                // exception for scheduler to catch
+                if (auto block_ex = cast(FiberBlockedException) t)
```

Comment:
If you cache the instance (assuming there's only one "in flight" at any time), you can also use `is` here, which does pointer comparison (as opposed to dynamic cast)